add covalency to badelf 🧝

SWeav02 opened the pull request:
@scott-materials 
This PR is in support of issue #727. The goal is to make it easier to run BadELF for users interested in electrides that contain covalency. 

So far, this PR does the following:

- Switch from "He" as our dummy atom to "X" for electrides and "Z" for covalent bonds
- Adds support to automatically catch covalent bonds and label them with "Z" atoms
- Allows for covalent bonds to be separated from nearby atoms with either a bader ("zero-flux") or plane ("voronoi") method
- Saves a list of the atoms that are involved in each covalent bonds to the database

I would also like to add the following features, (though maybe not in this PR):

- [ ] Change default small pseudopotentials to use larger GW potentials with more atoms, per discussion in #720 
- [x] Make covalent bond search more rigorous (e.g. include metal bonds)
- [ ] Automatically split charge from covalent bonds and assign to bonding atoms

**or**
- [ ] Provide better docs for meaning of oxidation states when covalent bonds are involved

The last two of these reflect different philosophies of how to handle covalent bonds. In the first, we would define our own method for splitting the charge on covalent bonds (e.g. split evenly, split at maxima, etc.). In the second we wouldn't split the charge at all and would leave interpretation up to the user. The first is more convenient for out of the box usage, but I think the second more closely reflects how the ELF community analyzes basins in molecules.

SWeav02 committed:
Add covalency finder

SWeav02 committed:
Add covalent alg options and docs

Andrew-S-Rosen commented on src/simmate/apps/badelf/core/badelf.py:
```diff
@@ -139,15 +193,18 @@ def _get_coord_envs(self):
+        # For each site in the structure, we add the coordination environment.
+
+        # We first need to replace any dummy atoms (X or Y) with an atom CrystalNN
+        # will recognize. We default to He since it is virtually never in a solid
```
Would picking something like `Og` be better here since VASP/QE calculations can't be run on it anyway? It probably doesn't matter, although for the inorganic electrides there is of course Na2He... 😅

SWeav02 replied:
> Maybe it would make sense to check if He is in `structure.composition` and raise an error if so?

Good point! Rather than raising an error, I can probably automatically find an atom not in the structure

SWeav02 replied:
Actually, I just realized it shouldn't matter which atom is used, even if it overlaps with what's in the structure. This is just a temporary structure object to get CrystalNN working and its not used anywhere else. Maybe there's a better atom than He to mimic the radius of an electride though? My first thought would be H-, although I think it tends to be a little smaller than electride sites.

Andrew-S-Rosen replied:
Hydride would certainly be the most natural choice, but I agree it will likely be smaller than for the electride site (size definition perhaps up for debate). If you think it might matter in meaningful ways for the algorithm, I suppose you can have a keyword argument defining the radius and then define a custom Pymatgen `Element` that is effectively H but with the user-defined radius used in the `CrystalNN` search. Might not be needed though...

Andrew-S-Rosen commented on src/simmate/apps/badelf/core/badelf.py:
```diff
@@ -139,15 +193,18 @@ def _get_coord_envs(self):
         # create a CrystalNN loop
         cnn = CrystalNN()
```
Small comment that is perhaps for another PR (which I can certainly tackle if you would like): I wonder if it would be beneficial to allow the user to customize the arguments in `CrystalNN` since the class tends to be fairly sensitive to things like the `x_diff_weight`. The question is how to do that in a clean way. Having `crystalnn_kwargs: dict | None = None` is fine but not particularly elegant, and I don't know if Simmate can handle instantiated classes as keyword arguments or not (e.g. if there will be serialization issues).

jacksund replied:
if you want this option exposed at the workflow level, your suggestion of `crystalnn_kwargs` is the way to go. Simmate can handle serialization of initialized classes, but I opt for `*_kwargs` style settings because it keeps `yaml` and `toml` workflow submissions viable. I do this in several places already -- for example: https://jacksund.github.io/simmate/parameters/#validator_kwargs

but like you said, it's not as elegant to those comfortable in python 🤷

That said, `BadElfToolkit` itself can accept an initialized class, while the higher-level workflow uses `*_kwargs`. It's all up to you and Sam

Andrew-S-Rosen replied:
> Simmate can handle serialization of initialized classes, but I opt for *_kwargs style settings because it keeps yaml and toml workflow submissions viable.

Cool! That definitely makes sense. All good to know, and I agree with that.

Of course, this is all Sam's decision. I am providing unsolicited feedback! 😅

SWeav02 replied:
Just a quick weigh-in, but I think this is probably a bit overkill unless there's interest in using coordination environments in a high-throughput way. CrystalNN isn't used in the main algorithm, it's just used here to give users an estimate of the coordination environment in the summary. I added this because @scott-materials and I found it made it easier to qualitatively think through some of our results.

When I find neighbors in the main algorithm, I consider all atoms within a large radius and then reduce from there (most of the method is [here](https://github.com/jacksund/simmate/blob/91af10b06c5bc37f045aa4d3b15d5453377affa0/src/simmate/apps/badelf/core/partitioning.py#L1033-L1202)). I found CrystalNN (and VoronoiNN) didn't always catch neighbors that ended up being important to the final partitioning surface.

Andrew-S-Rosen commented on src/simmate/apps/badelf/core/electride_finder.py:
```diff
@@ -168,14 +168,14 @@ def get_electride_structure(
                 calculation.
 
-            check_for_covalency (bool):
+            ignore_low_pseudopotentials (bool):
                 Whether to prevent a structure from being found if covalency is
                 found in the structure. It is highly recommended to keep this as
                 True as there is currently no method implemented to handle
                 covalency.
 
```
Same comment here about how "It is highly recommended to keep this as True..." should probably read `False` now.

SWeav02 commented on docs/parameters.md:
```diff
@@ -436,6 +435,24 @@ For evolutionary searches, this is the value that should be optimized. Specifica
 --------------------------
 
+## ignore_low_pseudopotentials
+This parameter is unique to the badelf workflows of the warrenapp. It indicates whether the algorithm should throw an exception when the used pseudopotential didn't contain enough valence electrons. It is generally recommended to leave this as True as the results may be nonsense.
```
Yes! Thanks for the catch.

SWeav02 commented on src/simmate/apps/badelf/core/badelf.py:
```diff
@@ -56,6 +56,7 @@ def __init__(
         threads: int = None,
         algorithm: str = "badelf",
         find_electrides: bool = True,
+        covalent_bond_alg: str = "zero-flux",  # other option is "voronoi"
```
Good idea! I didn't know about the Literal type hint in Python.

SWeav02 committed:
Tutorial tweeks and H-

SWeav02 commented:
@scott-materials @Andrew-S-Rosen 

I haven't made any commits in a while, but I'm still working on this PR. I'm trying to rework the `ElectrideFinder` class to categorize features in the ELF into atoms, lone pairs, covalent bonds, metal bonds, and electrides. I'm hoping to implement some ideas from traditional ELF topology analysis like what's described here: https://www.degruyter.com/document/doi/10.1524/zkri.220.5.399.65073/html?lang=en&srsltid=AfmBOoq1cg24bxUImk1C3O9Fu7yFk2BvfgAxTpVOkVrV7H2OK7pGocV2

I'm not sure how long this will take, but wanted to let everyone know.

SWeav02 committed:
Update ElectrideFinder

SWeav02 commented:
@scott-materials @Andrew-S-Rosen 

I just made a commit that rebuilds the ElectrideFinder class from scratch :rocket: 
I've switched to a more rigorous method for finding electride sites based on traditional ELF topology analysis. The method can automatically distinguish atom cores, atom shells, covalent bonds, metallic features, and bare electrons. I believe this should work for molecules and molecular solids as well, though I haven't tested it extensively yet. As part of this update, I added a number of options for the user to have some control over what is considered an electride site. I also added the ability to treat the spin-up and spin-down ELF separately. Over the next couple of days I plan to build out the docs/tutorials with information on how the new ElectrideFinder works. This will go into more detail on how each ELF feature is assigned and how to use it outside the context of BadELF if you're interested.

Before I merge this PR there are a couple other things I'd like to work on. First, I want to update the BadELfToolkit class to be able to handle the spin-up and spin-down ELF/Charge density separately like described in issue #713. During the last couple weeks I've realized that's extremely important for many electrides. Second, I may try and add a method for assigning charge on the covalent/metallic bonds to nearby atoms. As it stands currently, nothing happens with these charges and atoms in covalent/metallic systems have nonsense oxidation states.

Sorry this took so long! It turns out ELF topology can get quite complex :laughing:. It was a ton of fun to deep dive into though!

Andrew-S-Rosen commented:
Very cool! Thanks for the update, @SWeav02. It's great to see the changes you were able to make here. CC'ing @JuliaBaratta to keep her in the loop on this thread based on her original inquiry. I'm sure Julia would be happy to test out a few cases too once the `ElectrideFinder` class is in good shape.

SWeav02 committed:
add spin and update models

SWeav02 committed:
Merge remote-tracking branch 'origin/main' into covalent_badelf

SWeav02 committed:
start updating docs

SWeav02 committed:
docs update

SWeav02 committed:
Finish Doc Text

SWeav02 committed:
Add images to docs

SWeav02 committed:
Parameters and final edits

SWeav02 committed:
Change name to ElfAnalyzer

SWeav02 committed:
Merge branch 'jacksund:main' into covalent_badelf

jacksund commented on docs/apps/badelf/finder/elf_analyzer:
you forgot the `.md` file ending on this one, so it doesn't open up when I test locally. Otherwise PR looks good to go!

SWeav02 committed:
add .md
